lib/AST/NameLookup.cpp

```diff
@@ -1646,9 +1646,18 @@ bool DeclContext::lookupQualified(Type type,
   // Handle nominal types.
   bool wantProtocolMembers = (options & NL_ProtocolMembers);
   bool wantLookupInAllClasses = false;
-  if (auto nominal = type->getAnyNominal()) {
-    visited.insert(nominal);
-    stack.push_back(nominal);
+  if (auto *NTD = type->getAnyNominal()) {
+    if (NTD->isGenericContext()) {
+      auto *DC = NTD->getDeclContext();
+      // Refuse to return results for situation when
+      // we have a nested type in generic context.
+      if (DC->getAsProtocolOrProtocolExtensionContext() ||
+          (DC->isLocalContext() && DC->isGenericContext()))
```

Comment:
In the past we've had a lot of crashes caused by "bail out" hacks like this. Can you fix the underlying issue instead of hacking around it? If you hack around it, you will probably just find a similar crash later. Eventually we might support types nested inside protocols, and it would be a pain to hunt down all the hacks put in to work around the underlying problems.

Reply:
As far as I understand to actually fix the problem would mean to implement support for nested types in generics because the crash happens in `openGeneric`'s `bindArchetypesFromContext` when it tries to access generic environment of the declaration context.

Reply:
We already support types nested inside generic types, just not protocols or local contexts. It shouldn't be hard to make openGeneric() do something self-consistent here. In fact I put in bindArchetypesFromContext() to fix crashes with types inside local contexts (even though the rest of the compiler doesn't support it yet). So fixing up that method would be preferable to putting in a hack that makes it unreachable altogether. If it becomes unreachable it will just bitrot and we'll have more debugging to catch up on when the rest of the compiler learns about types in local contexts.

Reply:
Ok, what I was going for originally was fixing `bindArchetypesFromContext` but I talked to @DougGregor  and it seemed like making lookup not return results in such case would be good idea. I'll do some more digging.